### PR TITLE
Keras: Simplify recurrent state updates

### PR DESCRIPTION
**Pull request by lgeiger**
This PR replaces the collection of state updates with list comprehensions to improve readability.

### CODE REVIEW

**Comment by lgeiger** [comment]
For some reason GitHub didn't accept my rebase the first time. Sorry for the noise and reviews I have requested due to my first push. I don't have permissions to remove them 😇

**Comment by rthadur** [comment]
@lgeiger can you please resolve conflicts ?

**Comment by lgeiger** [comment]
> @lgeiger can you please resolve conflicts ?

@rthadur I rebased this a month ago (right after you asked me to), and now again because it accumulated merge conflicts in the mean time. If you are interested in merging this please go ahead or close it otherwise.

**Comment by rthadur** [comment]
sure , @tanzhenyu could you please review

**Comment by tanzhenyu** [comment]
Reassigning to someone more familiar with recurrent layers.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by tanzhenyu — DISMISSED** [review]

**Comment by rthadur** [comment]
@lgeiger can you please check build failures ?

**Comment by lgeiger** [comment]
> @lgeiger can you please check build failures ?

@rthadur The windows build failures seams unrelated and the `pylint` task fails with `tensorflow/python/keras/layers/serialization.py:32: [W0622(redefined-builtin), ] Redefining built-in 'zip'` in a file that isn't touched by this PR.

I can rebase, if the failures are fixed on master, but that would require an additional approval to re-run CI.

**Comment by rthadur** [comment]
@lgeiger sure please make changes i can start CI.

**Comment by lgeiger** [comment]
> @lgeiger sure please make changes i can start CI.

👍Rebased it

**Comment by rthadur** [comment]
@lgeiger here are latest  sanity builds failures https://source.cloud.google.com/results/invocations/a242630c-9f53-4fb6-be89-f7e4bcd83a38/log , please check

**Comment by lgeiger** [comment]
@rthadur thanks for the link. I had to disable the pylint rule in 1dfd5fb17cdb88edb39a408b6093e80b5b7c25c1

File: tensorflow/python/keras/layers/convolutional_recurrent.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 from __future__ import print_function
 
 import numpy as np
+from six.moves import zip  # pylint: disable=redefined-builtin
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional_recurrent.py]: why is this needed?

Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_recurrent.py]: The behaviour of `zip` changed in Python 3. To ensure the same behaviour, in Python 2 it will use `itertools.izip()` where as in Python 3 it will use `zip()`. For more info see: https://six.readthedocs.io/#module-six.moves

Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_recurrent.py]: If it is fine that `zip` returns a `list` in Python 2 and an `iterator` in Python 3, I can switch to using the builtin `zip` here.

Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_recurrent.py]: @tanzhenyu With TensorFlow slowly dropping support for Python 2, this minor detail doesn't matter anymore, so I removed the import from the PR.

**Review by tanzhenyu — APPROVED** [review]
